### ISSUE TITLE
How to pass stdio to nodejs

### ISSUE THREAD

**Issue by soroshsabz** [JeringTech/Javascript.NodeJS#163]
ITNOA

I want to link my stdio from C# Console Application to NodeJs, to better describe my question, I create a sample like below

I have some code like below

```
            IJsEngine engine = new NodeJsEngine(new NodeSettings { UseBuiltinLibrary = true});
            engine.ExecuteFile("SpawnServer.js");
```

and my `SpawnServer.js` is like below

```
const { spawn } = require("child_process");

console.log('Hi');
spawn("npx", ["ansible-language-server"].concat(" --stdio"), { stdio: 'inherit' });
```

I like to see `Hi` in my C# Console Application, But I do not know, How to do it?

**Note: All code is in https://github.com/soroshsabz/visualstudio-ansible/tree/features/4-support-ansible-language-server**

thanks

**Comment by soroshsabz** [comment]
I add this question in [SO](https://stackoverflow.com/q/75806882/1539100)

**Comment by soroshsabz** [comment]
@Taritsyn Did you have any idea?

**Comment by soroshsabz** [comment]
@JeremyTCD Did you have any Idea? or if this library does not support this, Did you have any advice to how to add this feature? maybe I can provide PR for this

**Comment by JeremyTCD** [comment]
Hey apologies for the slow response, I'm not sure how to pipe output when using JavaScriptEngineSwitcher, but if you use this library directly, you can do it like this:

```cs
using Jering.Javascript.NodeJS;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace Example.Logging
{
    internal class Program
    {
        static async Task Main(string[] _)
        {
            // Prepare DI services
            var serviceCollection = new ServiceCollection();
            serviceCollection.AddNodeJS();
            serviceCollection.AddLogging((ILoggingBuilder loggingBuilder) =>
            {
                loggingBuilder.AddConsole();
            });

            // Get INodeJSService
            IServiceProvider serviceProvider = serviceCollection.BuildServiceProvider();
            INodeJSService nodeJSService = serviceProvider.GetRequiredService<INodeJSService>();

            // Invoke js
            await nodeJSService.InvokeFromStringAsync(@"module.exports = (callback) => { 
    console.log('Hello world!'); 
    callback(null); 
}").ConfigureAwait(false);
        }
    }
}
```

Your project will have to reference the `Jering.Javascript.NodeJS` and `Microsoft.Extensions.Logging.Console` packages.

You will get output like:

```
info: Jering.Javascript.NodeJS.HttpNodeJSService[0]
      Connected to NodeJS through HTTP/1.1. Endpoint: http://[::1]:62220/.
info: Jering.Javascript.NodeJS.HttpNodeJSService[0]
      Connected to NodeJS process: 11960.
info: System.Net.Http.HttpClient.IHttpClientService.LogicalHandler[100]
      Start processing HTTP request POST http://[::1]:62220/
info: System.Net.Http.HttpClient.IHttpClientService.ClientHandler[100]
      Sending HTTP request POST http://[::1]:62220/
info: Jering.Javascript.NodeJS.HttpNodeJSService[0]
      Hello world!
info: System.Net.Http.HttpClient.IHttpClientService.ClientHandler[101]
      Received HTTP response headers after 101.8191ms - 200
info: System.Net.Http.HttpClient.IHttpClientService.LogicalHandler[101]
      End processing HTTP request after 108.5691ms - 200
```

You can filter output by following Microsoft's instructions [here](https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#non-host-console-app).

Let me know if you have any issues.

**Comment by soroshsabz** [comment]
@JeremyTCD are you sure, this method, redirect standard I/O to/from NodeJS, for example, what is happen, if I run below code in the NodeJS

```js
const stdin = process.openStdin()

process.stdout.write('Enter name: ')

stdin.addListener('data', text => {
  const name = text.toString().trim()
  console.log('Your name is: ' + name)

  stdin.pause() // stop reading
})
```

Are you sure, nodeJs get input from my standard input?

thanks

**Comment by JeremyTCD** [comment]
Ah it does not. 

I assumed from your "hi" example that you just needed output, stdin is not piped from .Net to Node.js.

Would an invocation work? E.g. 

```cs
string? data = Console.ReadLine();
nodeJSService.InvokeFromStringAsync(..., args: new object?[] { data });
```

**Comment by soroshsabz** [comment]
@JeremyTCD No, I want to bind (redirect) all input and output together, from/to .Net Node.js. for example, I have console application in Node.JS and I want when run this app in .Net, user can seamless using this app.

thanks

**Comment by JeremyTCD** [comment]
Options:

- Use invocations to "pipe" input to the Node.js process.
- Register a custom service for `[NodeJSProcessFactory](https://github.com/JeringTech/Javascript.NodeJS/blob/master/src/NodeJS/NodeJSServiceImplementations/OutOfProcess/NodeJSProcessFactory.cs)` that pipes input to the Node.js process.

**Comment by JeremyTCD** [comment]
Options:

- Use invocations to "pipe" input to the Node.js process.
- Register a custom service for `[NodeJSProcessFactory](https://github.com/JeringTech/Javascript.NodeJS/blob/master/src/NodeJS/NodeJSServiceImplementations/OutOfProcess/NodeJSProcessFactory.cs)` that pipes input to the Node.js process.

**Comment by soroshsabz** [comment]
I do not understand, did you can provide some example? (If I understand correctly, you says Jering can support redirection both standard output and input? correct?)

**Comment by JeremyTCD** [comment]
Piping input to Node.js processes is not supported out-of-the-box. But there are ways to implement such behaviour.

For the first option, read input and pass it on to the Node.js process using invocations, like in [this comment](https://github.com/JeringTech/Javascript.NodeJS/issues/163#issuecomment-1484518926).

As for the second option, replace the  `INodeJSProcessFactory` service that creates Node.js processes with a custom one that pipes input to them. Replacement of the service is done using standard [Microsoft DI](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection) patterns. You can the use [existing service](https://github.com/JeringTech/Javascript.NodeJS/blob/master/src/NodeJS/NodeJSServiceImplementations/OutOfProcess/NodeJSProcessFactory.csy) to get started. [Here in Microsoft's docs there is an example on piping to a process' stdin](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standardinput?view=net-7.0).

**Comment by janpaweljanusz** [comment]
@JeremyTCD Could you please give one more hint on how it can be achieved? I work on two directional communication as well and currently I can't figure it out. I would much appreciate any help.
